proposals/runtime_get_contexts.md

```diff
@@ -0,0 +1,259 @@
+`extension.getViews()`.
+
+## Problem
+
+With the migration to Manifest V3 (MV3), background pages
+[no longer exist](https://developer.chrome.com/docs/extensions/mv3/migrating_to_service_workers/);
+instead, an extension's background context is
+[service worker](https://developer.chrome.com/docs/workbox/service-worker-overview/)-based.
+For technical reasons<sup>[1](#footnotes)</sup>, service workers cannot access
+the [`Window`](http://go/mdn/API/Window#instance_properties) objects that
+`extension.getViews()` provides and it is not feasible to implement that with
+our browser design. Due to this, we cannot provide access to the JavaScript
+context for these views, but we can allow an extension to query for them
+(determining if they exist) and target them for messaging purposes.
+
+## Solution
+
+Considering the above situation, we'd like to propose a new extension API
+method, `runtime.getContexts()`, to asynchronously provide metadata about
+associated contexts that is still useful for an extension. A "context" here is
+considered an environment running the extension's code; the contexts we will
+consider are described in more detail in later sections. This will allow
+extension background scripts to identify the active contexts in the extension.
+For example, this can be used to target messages to send using
+[`runtime.sendMessage()`](https://developer.chrome.com/docs/extensions/reference/runtime/#method-sendMessage),
+etc.). Introducing this API will allow an easier migration from MV2.
+
+This also obviates the need for introducing multiple one-off APIs, such as
+separate APIs to query for the extension popup, offscreen documents, etc.
+
+## API Proposal
+
+This method will return an array of matching contexts, represented by a new
+`ExtensionContext` type. This will be defined as:
+
+```js
+runtime.ExtensionContext = {
+  // Context type -- tab, popup, etc.
+  contextType: ContextType,
+  // A unique identifier for this context.
+  contextId: string,
+  // ID of the tab for this context, or -1 if this context is not hosted in a
+  // tab.
+  tabId: int,
+  // ID of the window for this context, or -1 if this context is not hosted in a
+  // window.
+  windowId: int,
+  // ID of the DOM document for this context, or undefined if this context is
+  // not associated with a document.
+  documentId?: string,
+  // ID of the frame for this context, or -1 if this context is not hosted in a
+  // frame.
+  frameId: int,
+  // The current URL of the document, or undefined if this context is not
+  // hosted in a document.
+  documentUrl?: string,
+  // The current origin of the document, or undefined if this context is not
+  // hosted in a document.
+  documentOrigin?: string,
+  // Whether the context is for an incognito profile.
+  incognito: boolean,
+}
+```
+
+`ContextType` will indicate the type of context retrieved. It is an enum
+defined as:
+
+```js
+extension.ContextType = {
```

Comment:
Is this enum meant to be exhaustive? Can other browsers extend it? For example, a number of browsers support [sidebars](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/user_interface/Sidebars). Some extensions for Firefox and Vivaldi currently struggle with messaging with these contexts (since sidebars do not have a meaningful `tabId` and `frameId` for messaging. Developers can create persistent ports, but it is ugly.

Reply:
As with all APIs discussed in this community group, it's non-binding, and browser vendors can deviate when it makes sense.  If a browser supports a different type of context, they can add it in here, and if they don't support a given context, they can remove the enum entry.

FWIW, Chrome is also going to support side panels, so we will add a SIDE_PANEL context type when that happens.  (I was tempted to include it here, but didn't want to confuse the proposal by having unshipped concepts).

Reply:
Thanks for the call-out!  Devtools are interesting -- in Chromium, they commit to a different origin (even though from the extension's POV, they are just the extension page).  They are also not currently included at all in extension.getViews().  I'd like to tackle these separately, but I've added an explicit section for them in future work.

I also added in a type for SIDE_PANEL (though it's still under development in Chrome)